clients/src/main/java/org/apache/kafka/common/metrics/MetricsReporter.java

```diff
@@ -65,4 +66,12 @@ default void validateReconfiguration(Map<String, ?> configs) throws ConfigExcept
     default void reconfigure(Map<String, ?> configs) {
     }
 
+    /**
+     * Provides context labels for the service or library exposing metrics
+     *
+     * @param metricsContext the metric context
+     */
+    @InterfaceStability.Evolving
+    default void contextChange(MetricsContext metricsContext) {
```

Comment:
someone on the ML commented that we might want to name this `contextChanged` (past tense). I don't have a strong feeling either way. Do you have any thoughts @mumrah @rhauch?

Reply:
I really don't have a preference. Past tense is a little odd, but I could see where `changeContext(...)` or `setContext(...)` are present tense and more conventional.

Reply:
It would also be good to identify when this is called relative to other methods. For example, it is always called before `init(...)` is called. But can it be called again, or is that the only time this method is called?

Reply:
due to the way jmxreporter is initialized in Kafka today, it already gets called both before and after `init()`

Reply:
It can be called multiple times.  Not sure if we should mention that in Javadoc, other methods in this class we are not mention if it can be called multiple times even though they can be called multiple times.

Reply:
If there is no clear call pattern, then it's fine to not say anything. However, `JmxReporter.contextChange(...)` seems to assume that `contextChange(...)` will be called before any metrics are added via `init(...)`.

If that call pattern is true, then I think we should document it. If it's also true it can be called later, then mention this as well. For example, the JavaDoc text on `contextChange(...)` could be something like:

> Sets the context labels for the service or library that is exposing metrics. 
> This will be called before {@link #init(List)} and may be called anytime after that.

WDYT?

Reply:
Sounds good. Updated javadoc.